src/block/MonsterSpawner.php

```diff
@@ -23,12 +23,102 @@
 
 namespace pocketmine\block;
 
+use pocketmine\block\tile\MonsterSpawner as TileSpawner;
+use pocketmine\block\tile\SpawnerSpawnRangeRegistry;
 use pocketmine\block\utils\SupportType;
+use pocketmine\entity\Entity;
+use pocketmine\entity\EntityFactory;
+use pocketmine\entity\Location;
+use pocketmine\event\block\SpawnerAttemptSpawnEvent;
 use pocketmine\item\Item;
+use pocketmine\item\SpawnEgg;
+use pocketmine\item\SpawnEggEntityRegistry;
+use pocketmine\math\AxisAlignedBB;
+use pocketmine\math\Vector3;
+use pocketmine\nbt\tag\CompoundTag;
+use pocketmine\nbt\tag\DoubleTag;
+use pocketmine\nbt\tag\FloatTag;
+use pocketmine\nbt\tag\ListTag;
+use pocketmine\player\Player;
+use pocketmine\world\particle\MobSpawnParticle;
+use function assert;
 use function mt_rand;
 
 class MonsterSpawner extends Transparent{
 
+	/** TODO: replace this with a cached entity or something of that nature */
+	private string $entityTypeId = TileSpawner::DEFAULT_ENTITY_TYPE_ID;
+	/** TODO: deserialize this properly and drop the NBT (PC and PE formats are different, just for fun) */
+	private ?ListTag $spawnPotentials = null;
+	/** TODO: deserialize this properly and drop the NBT (PC and PE formats are different, just for fun) */
+	private ?CompoundTag $spawnData = null;
+
+	private float $displayEntityWidth = 1.0;
+	private float $displayEntityHeight = 1.0;
+	private float $displayEntityScale = 1.0;
+
+	private int $spawnDelay = TileSpawner::DEFAULT_MIN_SPAWN_DELAY;
+	private int $minSpawnDelay = TileSpawner::DEFAULT_MIN_SPAWN_DELAY;
+	private int $maxSpawnDelay = TileSpawner::DEFAULT_MAX_SPAWN_DELAY;
+	private int $spawnPerAttempt = 1;
+	private int $maxNearbyEntities = TileSpawner::DEFAULT_MAX_NEARBY_ENTITIES;
+	private int $spawnRange = TileSpawner::DEFAULT_SPAWN_RANGE;
+	private int $requiredPlayerRange = TileSpawner::DEFAULT_REQUIRED_PLAYER_RANGE;
+
+	public function readStateFromWorld() : Block{
+		parent::readStateFromWorld();
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if($tile instanceof TileSpawner){
+			$this->entityTypeId = $tile->getEntityTypeId();
+			$this->spawnPotentials = $tile->getSpawnPotentials();
+			$this->spawnData = $tile->getSpawnData();
+			$this->displayEntityWidth = $tile->getDisplayEntityWidth();
+			$this->displayEntityHeight = $tile->getDisplayEntityHeight();
+			$this->displayEntityScale = $tile->getDisplayEntityScale();
+			$this->spawnDelay = $tile->getSpawnDelay();
+			$this->minSpawnDelay = $tile->getMinSpawnDelay();
+			$this->maxSpawnDelay = $tile->getMaxSpawnDelay();
+			$this->spawnPerAttempt = $tile->getSpawnPerAttempt();
+			$this->maxNearbyEntities = $tile->getMaxNearbyEntities();
+			$this->spawnRange = $tile->getSpawnRange();
+			$this->requiredPlayerRange = $tile->getRequiredPlayerRange();
```

Comment:
Not liking this. A data class to wrap all these up into one item to be passed between block & tile would be a lot cleaner.